compose/project.py

```diff
@@ -586,8 +586,10 @@ def _get_convergence_plans(self, services, strategy, always_recreate_deps=False)
                           ", ".join(updated_dependencies))
                 containers_stopped = any(
                     service.containers(stopped=True, filters={'status': ['created', 'exited']}))
-                has_links = any(c.get('HostConfig.Links') for c in service.containers())
-                if always_recreate_deps or containers_stopped or not has_links:
+                service_has_links = any(service.get_link_names())
+                container_has_links = any(c.get('HostConfig.Links') for c in service.containers())
+                should_recreate_for_links = service_has_links ^ container_has_links
```

Comment:
It took me a while to puzzle out this xor, essentially it is reconcilling what the service asks for (which may have changed) with what the running container has.I think my main mental barrier was the commit message saying "if the service defines links but the container does not have any" which is only half the story, since we also should reconcille if the service has no links but the container does have them. If that's correct would you mind updating the commit message to reflect that please.

Reply:
I think the tests don't actually cover the case of adding / removing links dynamically (only close one is `test_service_removed_while_down` but the project is down for that change). Is that also right?

Reply:
i have added 2 new tests `test_dependency_removed` and `test_dependency_added` which should explicitly verify the adding/removing links dynamically behaves as expected

Reply:
have also modified the commit message as requested. hopefully addresses your comment